concurrency/CreativeServiceLiveConfigImpl.java

```diff
@@ -0,0 +1,182 @@
+    }
+    
+    void loadCreativeTemplates(long bidderId, Set<CreativeTemplate> templates) {
+        Map<String, Set<CreativeTemplate>> templatesPerSize = new HashMap<String, Set<CreativeTemplate>>();
+        Map<String, CreativeTemplate> templatesById = new HashMap<String, CreativeTemplate>();
+        for (CreativeTemplate template : templates) {
+            Set<CreativeTemplate> sizeTemplates = templatesPerSize.get(template.getSize());
+            if (sizeTemplates == null) {
+                sizeTemplates = new HashSet<CreativeTemplate>();
+                templatesPerSize.put(template.getSize(), sizeTemplates);
+            }
+            sizeTemplates.add(template);
+            templatesById.put(template.getTemplateId(), template);
+        }
+        
+        bidderTemplatesBySize.put(bidderId, templatesPerSize);
+        bidderTemplatesById.put(bidderId, templatesById);
+        bidderCompiledTemplates.remove(bidderId);
+    }
+
+    @Override
+    public CreativeTemplate getTemplate(long bidderId, String templateId) {
+        Set<CreativeTemplate> templates = liveConfig.getCreativeTemplatesForBidder(bidderId);
+        if (CollectionUtils.isEmpty(templates)) return null;
+        
+        try {
+            Set<CreativeTemplate> cachedTemplates = bidderTemplates.get(bidderId);
+            if (cachedTemplates == null || !templates.equals(cachedTemplates)) {
+                loadCreativeTemplates(bidderId, templates);
+            }
+            
+            return bidderTemplatesById.get(bidderId).get(templateId);
+        } catch (Exception e) {
+            PMETUtil.increment(aaxLiveConfigCreativeServiceErrorCounter);
+            logger.warn("Error in getTemplate; bidderId=%d, templateId=%s", e, bidderId, templateId);
+        }
+        
+        return null;
+    }
+
+    @Override
+    public StringTemplate getCompiledTemplate(long bidderId, String templateId) {
+        StringTemplate compiledTemplate;
+        try {
+            compiledTemplate = bidderCompiledTemplates.get(bidderId).get(templateId);
+            if (compiledTemplate != null) return compiledTemplate;
+        } catch (Exception e) {
+            // ignore
+        }
+        
+        CreativeTemplate creativeTemplate = getTemplate(bidderId, templateId);
+        if (creativeTemplate == null) return null;
+        
+        try {
+            compiledTemplate = new StringTemplate(creativeTemplate.getHtml());
+            ConcurrentHashMap<String, StringTemplate> stringTemplates = bidderCompiledTemplates.get(bidderId);
+            if (stringTemplates == null) {
+                synchronized(bidderCompiledTemplates) {
+                    stringTemplates = bidderCompiledTemplates.get(bidderId);
+                    if (stringTemplates == null) {
+                        stringTemplates = new ConcurrentHashMap<String, StringTemplate>();
+                        bidderCompiledTemplates.put(bidderId, stringTemplates);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.



**Problem**  
 These lines use `ConcurrentHashMap`&#39;s **sequential** API and are protected by `synchronized` to ensure they execute together: **152 and 155**. This synchronization may reduce performance by limiting parallelism and does not take advantage of `ConcurrentHashMap`&#39;s **concurrent** API.

**Fix**  
 Consider replacing the check and `put()` with `putIfAbsent()` and then removing the synchronization. The check is at line: **153**.

**More info**  
 `putIfAbsent()` is thread-safe and puts the value only if the `ConcurrentHashMap` does not contain the key. `putIfAbsent()` returns `null` if the value did not exist and returns the value in the map if one already exists.

 [View an example on GitHub](https://github.com/apache/hadoop/blob/f16cf877e565084c66bc63605659b157c4394dc8/hadoop-mapreduce-project/hadoop-mapreduce-client/hadoop-mapreduce-client-hs/src/main/java/org/apache/hadoop/mapreduce/v2/hs/HistoryFileManager.java#L927-L932) (external link).